diff --git a/Dockerfile b/Dockerfile
index 2ffaf4a815..9049b67055 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -15,11 +15,13 @@ WORKDIR /influxdb3
 ARG CARGO_INCREMENTAL=yes
 ARG CARGO_NET_GIT_FETCH_WITH_CLI=false
 ARG PROFILE=release
-ARG FEATURES=aws,gcp,azure,jemalloc_replacing_malloc
+#ARG FEATURES=aws,gcp,azure,jemalloc_replacing_malloc
+ARG FEATURES=aws,jemalloc_replacing_malloc
 ARG PACKAGE=influxdb3
 ARG PBS_DATE=unset
 ARG PBS_VERSION=unset
 ARG PBS_TARGET=unset
+ARG TARGETARCH
 ENV CARGO_INCREMENTAL=$CARGO_INCREMENTAL \
     CARGO_NET_GIT_FETCH_WITH_CLI=$CARGO_NET_GIT_FETCH_WITH_CLI \
     PROFILE=$PROFILE \
@@ -27,11 +29,18 @@ ENV CARGO_INCREMENTAL=$CARGO_INCREMENTAL \
     PACKAGE=$PACKAGE \
     PBS_TARGET=$PBS_TARGET \
     PBS_DATE=$PBS_DATE \
+    JEMALLOC_SYS_WITH_LG_PAGE=16 \
+    CARGO_BUILD_JOBS=1 \
     PBS_VERSION=$PBS_VERSION
 
+RUN case "${TARGETARCH}" in \
+        "amd64")  export PBS_TARGET="x86_64-unknown-linux-gnu" ;; \
+        "arm64") export PBS_TARGET="aarch64-unknown-linux-gnu" ;; \
+        *) echo "unsupported architecture ${TARGETARCH}" && exit 1 ;; \
+    esac && echo "PBS_TARGET=${PBS_TARGET}" > /tmp/pbs.properties
 # obtain python-build-standalone and configure PYO3_CONFIG_FILE
 COPY .circleci /influxdb3/.circleci
-RUN \
+RUN . /tmp/pbs.properties && \
   sed -i "s/^readonly TARGETS=.*/readonly TARGETS=${PBS_TARGET}/" ./.circleci/scripts/fetch-python-standalone.bash && \
   ./.circleci/scripts/fetch-python-standalone.bash /influxdb3/python-artifacts "${PBS_DATE}" "${PBS_VERSION}" && \
   tar -C /influxdb3/python-artifacts -zxf /influxdb3/python-artifacts/all.tar.gz "./${PBS_TARGET}" && \
@@ -39,20 +48,12 @@ RUN \
   cat "/influxdb3/python-artifacts/${PBS_TARGET}/pyo3_config_file.txt"
 
 COPY . /influxdb3
+#COPY corporateproxy.crt /usr/local/share/ca-certificates/
+RUN update-ca-certificates
 
-RUN \
-  --mount=type=cache,id=influxdb3_rustup,sharing=locked,target=/usr/local/rustup \
-  --mount=type=cache,id=influxdb3_registry,sharing=locked,target=/usr/local/cargo/registry \
-  --mount=type=cache,id=influxdb3_git,sharing=locked,target=/usr/local/cargo/git \
-    du -cshx /usr/local/rustup /usr/local/cargo/registry /usr/local/cargo/git && \
-    rustup toolchain install
-
-RUN \
-  --mount=type=cache,id=influxdb3_rustup,sharing=locked,target=/usr/local/rustup \
-  --mount=type=cache,id=influxdb3_registry,sharing=locked,target=/usr/local/cargo/registry \
-  --mount=type=cache,id=influxdb3_git,sharing=locked,target=/usr/local/cargo/git \
-  --mount=type=cache,id=influxdb3_target,sharing=locked,target=/influxdb3/target \
-    du -cshx /usr/local/rustup /usr/local/cargo/registry /usr/local/cargo/git /influxdb3/target && \
+RUN rustup toolchain install
+
+RUN . /tmp/pbs.properties && \
     PYO3_CONFIG_FILE="/influxdb3/python-artifacts/$PBS_TARGET/pyo3_config_file.txt" cargo build --target-dir /influxdb3/target --package="$PACKAGE" --profile="$PROFILE" --no-default-features --features="$FEATURES" && \
     objcopy --compress-debug-sections "target/$PROFILE/$PACKAGE" && \
     cp "/influxdb3/target/$PROFILE/$PACKAGE" "/root/$PACKAGE" && \
