// @influxdata/vis/src/constants/colorSchemes.ts
export const VIRIDIS = [
  '#440154',
  '#481f70',
  '#443983',
  '#3b528b',
  '#31688e',
  '#287c8e',
  '#21918c',
  '#20a486',
  '#35b779',
  '#5ec962',
  '#90d743',
  '#c8e020',
]

export const MAGMA = [
  '#000004',
  '#100b2d',
  '#2c115f',
  '#51127c',
  '#721f81',
  '#932b80',
  '#b73779',
  '#d8456c',
  '#f1605d',
  '#fc8961',
  '#feb078',
  '#fed799',
]

export const INFERNO = [
  '#000004',
  '#110a30',
  '#320a5e',
  '#57106e',
  '#781c6d',
  '#9a2865',
  '#bc3754',
  '#d84c3e',
  '#ed6925',
  '#f98e09',
  '#fbb61a',
  '#f4df53',
]

export const PLASMA = [
  '#0d0887',
  '#3a049a',
  '#5c01a6',
  '#7e03a8',
  '#9c179e',
  '#b52f8c',
  '#cc4778',
  '#de5f65',
  '#ed7953',
  '#f89540',
  '#fdb42f',
  '#fbd524',
]

export const YL_OR_RD = [
  '#ffffcc',
  '#ffeda0',
  '#fed976',
  '#feb24c',
  '#fd8d3c',
  '#fc4e2a',
  '#e31a1c',
  '#bd0026',
  '#800026',
]

export const YL_GN_BU = [
  '#ffffd9',
  '#edf8b1',
  '#c7e9b4',
  '#7fcdbb',
  '#41b6c4',
  '#1d91c0',
  '#225ea8',
  '#253494',
  '#081d58',
]

export const BU_GN = [
  '#f7fcfd',
  '#ebf7fa',
  '#dcf2f2',
  '#c8eae4',
  '#aadfd2',
  '#88d1bc',
  '#68c2a3',
  '#4eb485',
  '#37a266',
  '#228c49',
  '#0d7635',
  '#025f27',
]

// Chronograf Colors
export const NINETEEN_EIGHTY_FOUR = ['#31C0F6', '#A500A5', '#FF7E27']
export const ATLANTIS = ['#74D495', '#3F3FBA', '#FF4D9E']
export const DO_ANDROIDS_DREAM = ['#8F8AF4', '#A51414', '#F4CF31']
export const DELOREAN = ['#FD7A5D', '#5F1CF2', '#4CE09A']
export const CTHULHU = ['#FDC44F', '#007C76', '#8983FF']
export const ECTOPLASM = ['#DA6FF1', '#00717A', '#ACFF76']
export const T_MAX_400_FILM = ['#F6F6F8', '#A4A8B6', '#545667']
